docs: add new openapi for issuer

lucacavallaro opened the pull request:
<!--- Please always add a PR description as if nobody knows anything about the context these changes come from. -->
<!--- Even if we are all from our internal team, we may not be on the same page. -->
<!--- Write this PR as you were contributing to a public OSS project, where nobody knows you and you have to earn their trust. -->
<!--- This will improve our projects in the long run! Thanks. -->

#### List of Changes

Proposal of stable APIs for Issuers

- `snake_case` 🎉
- `Product` is now `Dossier`
-  `Clause` now have a `type` field that can be `unfair`, `required` or `optional`
-  New timestamps on `Signature Request`
- New endpoint `POST /signers` to get `Signer` objects by `fiscal code`
- The issuer now can specify `coordinates`, `page` and `size` in order to create a new `signature field` into the PDF document

<!--- Describe your changes in detail -->

#### Motivation and Context

<!--- Why is this change required? What problem does it solve? -->

#### How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->

#### Screenshots (if appropriate):

#### Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [x] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

lucacavallaro committed:
docs: add new openapi for issuer

grausof commented on docs/openapi-issuer.yaml:
```diff
@@ -0,0 +1,570 @@
+          items:
+            oneOf:
+              - $ref: "#/components/schemas/SignatureField"
+              - $ref: "#/components/schemas/SignatureFieldToBeCreated"
```
```suggestion
          items:
            $ref: "#/components/schemas/SignatureField"
```

grausof commented on docs/openapi-issuer.yaml:
```diff
@@ -0,0 +1,570 @@
+          properties:
+            w: { type: number, minimum: 0 }
+            h: { type: number, minimum: 0 }
+        clause:
+          $ref: "#/components/schemas/Clause"
```
```suggestion
```

grausof commented on docs/openapi-issuer.yaml:
```diff
@@ -0,0 +1,570 @@
+
+    SignatureField:
+      type: object
+      properties:
+        unique_name:
+          type: string
+        clause:
+          $ref: "#/components/schemas/Clause"
+      required:
+        - unique_name
+        - clause
```
```suggestion
      properties:
        field_detail:
          oneOf: 
            - $ref: "#/components/schemas/FieldToBeCreated"
            - $ref: "#/components/schemas/FieldUniqueName"
        clause:
          $ref: "#/components/schemas/Clause"
      required:
        - field_detail
        - clause
```

grausof replied:
I prefer to have all the information relating to a field (coordinates, dimensions and page) in a single object such as signature_field_detail

lucacavallaro replied:
ok! I'll try to group them in order to make these properties easier to use

grausof commented on docs/openapi-issuer.yaml:
```diff
@@ -0,0 +1,570 @@
+        documents:
+          type: array
+          items:
+            oneOf:
```
```suggestion
            anyOf:
```

lucacavallaro replied:
why `anyOf`?

grausof replied:
`documents` can match one or more subschemas. There can be DocumentToBeUploaded and also DocumentUploaded in the same array

lucacavallaro replied:
yes, but that is the purpose of `oneOf` (if I understood correctly)

for each element it checks if the shape matches with one of the scheme provided

`anyOf` checks the shape of an object between one and more schemes

grausof replied:
Mhh... yes you are right!

gunzip replied:
anyOf is not supported by the codegen

hevelius reviewed (commented):
LGTM / I don't know perfectly the specs for issuer but I follow the comments of @grausof

gunzip commented on docs/openapi-issuer.yaml:
```diff
@@ -0,0 +1,570 @@
+            application/json:
+              schema:
+                $ref: "#/components/schemas/DossierDetailView"
+        "400":
```
All API that pass though the API management must handle: 400, 401, 403, 429.

see https://github.com/pagopa/io-functions-services/blob/master/openapi/index.yaml#L296

gunzip commented on docs/openapi-issuer.yaml:
```diff
@@ -0,0 +1,570 @@
+
+  /signers:
+    post:
+      operationId: getSignerByFiscalCode
```
why is this a post and not a get?

lucacavallaro replied:
in order to avoid passing the fiscal code as route param

grausof replied:
What's the problem with passing the fiscalCode as route param?

lucacavallaro replied:
because the fiscal code is sensitive data (pii) and I would like to avoid "involuntary logging"

gunzip commented on docs/openapi-issuer.yaml:
```diff
@@ -0,0 +1,570 @@
+            application/json:
+              schema:
+                $ref: "#/components/schemas/SignerDetailView"
+        "400":
```
what about 404?

gunzip commented on docs/openapi-issuer.yaml:
```diff
@@ -0,0 +1,570 @@
+            $ref: "#/components/schemas/Id"
+      responses:
+        "204":
+          description: Signature Request status successfully setted
```
```suggestion
          description: Signature Request status successfully set
```

gunzip commented on docs/openapi-issuer.yaml:
```diff
@@ -0,0 +1,570 @@
+          type: string
+        type:
+          type: string
+          enum: ["REQUIRED", "UNFAIR", "OPTIONAL"]
```
```suggestion
          x-extensible-enum: ["REQUIRED", "UNFAIR", "OPTIONAL"]
```

gunzip replied:
plase add documentation for these fields

lucacavallaro replied:
why `x-extensibile-enum` ?

gunzip replied:
we can use enum if we're absolutely certain that this list won't change in future.

lucacavallaro committed:
docs: update openapi-issuer

- document 429 & 401 error responses
- update shape of "signature field" resource
- fix typo

lucacavallaro committed:
docs: add timestamps to "dossier" resource

grausof commented on docs/openapi-issuer.yaml:
```diff
@@ -0,0 +1,627 @@
+        default:
+          $ref: "#/components/responses/Unexpected"
+
+  /signature-requests/{id}/status:
+    put:
+      operationId: putSignatureRequestStatus
+      tags:
+        - Signature Request
+      summary: Set the status of a Signature Request
+      parameters:
+        - in: path
+          name: id
+          required: true
+          schema:
+            $ref: "#/components/schemas/Id"
```
there is no requestBody. How can I specify the status?

lucacavallaro replied:
good catch! :D

grausof commented on docs/openapi-issuer.yaml:
```diff
@@ -0,0 +1,627 @@
+      tags:
+        - Signature Request
+      summary: Get the Upload Url for the specified document
+      parameters:
+        - in: path
+          name: req_id
+          required: true
+          schema:
+            $ref: "#/components/schemas/Id"
+        - in: path
+          name: doc_id
+          required: true
+          schema:
+            $ref: "#/components/schemas/Id"
```
Isn't it better to specify?
```
req_id -> signature_request_id
doc_id -> document_id
```

lucacavallaro replied:
Given the rest of the used path, I think their "meaning" is clear

lucacavallaro committed:
docs: fix openapi-issuer

add body to putSignatureRequestStatus

grausof commented:
⚠️ `gen-api-models` crashes using these specifications
```
UploadUrl -> src/infra/api-models/UploadUrl.ts
Error: SyntaxError: Expression expected. (59:7)
  57 |       
  58 |       
> 59 |       });
     |       ^
  60 | 
  61 |       export const SignatureField = t.exact(
  62 |         t.intersection(
error Command failed with exit code 1.
```

grausof commented on docs/openapi-issuer.yaml:
```diff
@@ -0,0 +1,635 @@
+    SignatureField:
+      type: object
+      properties:
+        attrs:
+          oneOf:
+            - $ref: "#/components/schemas/SignatureFieldAttributes"
+            - $ref: "#/components/schemas/SignatureFieldToBeCreatedAttributes"
```
```suggestion
        attrs:
          type: object
          oneOf:
            - $ref: "#/components/schemas/SignatureFieldAttributes"
            - $ref: "#/components/schemas/SignatureFieldToBeCreatedAttributes"
```

grausof commented on docs/openapi-issuer.yaml:
```diff
@@ -0,0 +1,635 @@
+      required:
+        - id
+
+    RequestSignatureBody:
+      type: object
+      properties:
+        dossier_id:
+          $ref: "#/components/schemas/Id"
+        signer_id:
+          allOf:
+            - $ref: "#/components/schemas/Id"
+            - example: "01BX5ZZKBKACTAV9WEVGEMMVS0"
+        expires_at:
+          $ref: "#/components/schemas/Timestamp"
+      required:
+        - dossier_id
+        - signer_id
```
Here we lack the forwarding mode whether by message or qrcode.

lucacavallaro replied:
yep

grausof replied:
we leave `send_message: true` ?

lucacavallaro replied:
`notify_user_via_io_message`

lucacavallaro committed:
docs: update openapi-issuer

add `notify_user_via_io_message` property

lucacavallaro committed:
docs: fix openapi spec

grausof commented on docs/openapi-issuer.yaml:
```diff
@@ -0,0 +1,639 @@
+
+    SignatureRequest:
+      type: object
+      properties:
+        id:
+          $ref: "#/components/schemas/Id"
+        dossier_id:
+          $ref: "#/components/schemas/Id"
+        expires_at:
+          $ref: "#/components/schemas/Timestamp"
```
`signer_id` is missing

lucacavallaro replied:
After giving it some thought, I decided it wasn't necessary to include `signer_id` in this response, maybe `dossier_id` is redundant either. Do you believe it should be included?

grausof replied:
I think it is useful as the `ISSUER` given a `signature_request_id` will no longer be able to trace the `signer_id` if he has not stored this information somewhere.

grausof commented on docs/openapi-issuer.yaml:
```diff
@@ -0,0 +1,639 @@
+      properties:
+        dossier_id:
+          $ref: "#/components/schemas/Id"
+        signer_id:
+          allOf:
+            - $ref: "#/components/schemas/Id"
+            - example: "01BX5ZZKBKACTAV9WEVGEMMVS0"
```
```suggestion
        signer_id:
          $ref: "#/components/schemas/Id"
```

grausof replied:
typo?

lucacavallaro replied:
nope: it a trick to provide an example with different ID

grausof replied:
I don't think this trick is necessary. The id is always generated on the server side so it cannot confuse the issuer. Furthermore, this type of structure does not make the generator work. 
CC @gunzip

grausof commented on docs/openapi-issuer.yaml:
```diff
@@ -0,0 +1,639 @@
+          $ref: "#/components/schemas/Timestamp"
+        documents:
+          type: array
+          items:
+            oneOf:
```
```suggestion
          items:
            type: array
            oneOf:
```

grausof commented on docs/openapi-issuer.yaml:
```diff
@@ -0,0 +1,639 @@
+            w: { type: number, minimum: 0 }
+            h: { type: number, minimum: 0 }
+
+    SignatureField:
+      type: object
+      properties:
+        attrs:
+          type: object
+          oneOf:
+            - $ref: "#/components/schemas/SignatureFieldAttributes"
+            - $ref: "#/components/schemas/SignatureFieldToBeCreatedAttributes"
```
```suggestion
    SignatureField:
      properties:
        attrs:
          oneOf:
            - $ref: "#/components/schemas/SignatureFieldAttributes"
            - $ref: "#/components/schemas/SignatureFieldToBeCreatedAttributes"
```
In this way the specifications are validated!

gunzip commented on docs/openapi-issuer.yaml:
```diff
@@ -0,0 +1,639 @@
+    description: production
+security:
+  - SubscriptionKey: []
+paths:
```
not urgent but you probably want to create "Links" sections with the returned URLs to store (upload) documents: https://swagger.io/docs/specification/links/

lucacavallaro committed:
docs: fix openapi-issuer

grausof reviewed (approved):
LGTM

lucacavallaro committed:
Merge branch 'main' into new-openapi-for-issuer

grausof commented on docs/openapi-issuer.yaml:
```diff
@@ -0,0 +1,639 @@
+        clause:
+          $ref: "#/components/schemas/Clause"
+      required:
+        - unique_name
```
`attrs` and not `unique_name`

lucacavallaro replied:
check out the latest spec https://github.com/pagopa/io-sign/blob/functions-refactor/apps/io-func-sign-issuer/openapi.yaml
